### PR TITLE
Add elastic scaling support in ParaInherent BenchBuilder 

### PR DESCRIPTION
**Pull request by sandreim**
Extracted Benchbuilder enhancements used in https://github.com/paritytech/polkadot-sdk/pull/3644 . Might still require some work to fully support all scenarios when disputing elastic scaling parachains, but it should be useful in writing elastic scaling runtime tests.

### CODE REVIEW

File: polkadot/runtime/parachains/src/builder.rs [hunk 1]
```diff
@@ -442,7 +449,7 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
 		self.validators = Some(validators_shuffled);
 		self.block_number = block_number;
 		self.session = target_session;
-		assert_eq!(paras::Pallet::<T>::parachains().len(), total_cores as usize);
+		// assert_eq!(paras::Pallet::<T>::parachains().len(), total_cores as usize);
```
Comment (alindima) [inline polkadot/runtime/parachains/src/builder.rs:452]: this needs to be removed, right?

File: polkadot/runtime/parachains/src/builder.rs [hunk 2]
```diff
@@ -476,16 +484,27 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
 			})
 			.collect();
 
+		let mut current_core_idx = 0u32;
+
 		for (seed, _) in concluding_cores.iter() {
```
Comment (alindima) [inline polkadot/runtime/parachains/src/builder.rs:489]: concluding_cores here is really concluding_paras

File: polkadot/runtime/parachains/src/builder.rs [hunk 3]
```diff
@@ -494,112 +513,144 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
...
-					.map(|val_idx| {
-						let public = validators.get(*val_idx).unwrap();
-						let sig = UncheckedSigned::<CompactStatement>::benchmark_sign(
-							public,
-							CompactStatement::Valid(candidate_hash),
-							&self.signing_context(),
-							*val_idx,
+
+				let para_id = ParaId::from(*seed);
+				prev_head = None;
```
Comment (alindima) [inline polkadot/runtime/parachains/src/builder.rs:536]: I think the `prev_head` declaration should be moved here, to make it clear that the value doesn't need to persist across different paras

File: polkadot/runtime/parachains/src/builder.rs [hunk 4]
```diff
@@ -624,7 +677,14 @@ impl<T: paras_inherent::Config> BenchBuilder<T> {
 					.cloned()
 					.unwrap_or(self.target_session);
 
-				let (para_id, core_idx, group_idx) = self.create_indexes(seed);
+				let para_id = ParaId::from(seed);
+				let core_idx = CoreIndex::from(current_core_idx);
+				current_core_idx +=1;
```
Comment (alindima) [inline polkadot/runtime/parachains/src/builder.rs:682]: the behaviour here is weird. `start..last` are now paraids. It'd be better if we had them represent cores and have this logic dispute specific cores (regardless of their paraids)

**Review by alindima — COMMENTED** [review]
I suggest we also add the benchmark from https://github.com/paritytech/polkadot-sdk/pull/3644. The current code on master should already allow supplying multiple candidates if elastic scaling mvp feature is enabled (but only one will be backed). We have a test that does that: https://github.com/paritytech/polkadot-sdk/blob/d55d4f64b50c858bce6fa1a429f50d486cc67d49/polkadot/runtime/parachains/src/inclusion/tests.rs#L1671

File: polkadot/runtime/parachains/src/builder.rs (hunk 1 above)
Comment (sandreim) [reply polkadot/runtime/parachains/src/builder.rs:452]: yeah, but maybe it can still be checked if there are no elastic paras.

File: polkadot/runtime/parachains/src/builder.rs (hunk 4 above)
Comment (sandreim) [reply polkadot/runtime/parachains/src/builder.rs:682]: There is more work to be done to support disputes on cores of elastic paras. Assuming elastic scaling not used, these are 1:1 mapped to cores. so it should be fine.

File: polkadot/runtime/parachains/src/builder.rs (hunk 1 above)
Comment (sandreim) [reply polkadot/runtime/parachains/src/builder.rs]: added back the correct check accounting for extra cores from elastic paras.

File: polkadot/runtime/parachains/src/builder.rs (hunk 3 above)
Comment (sandreim) [reply polkadot/runtime/parachains/src/builder.rs:541]: yes, that was moved around and forgot to remove this line.

Comment (sandreim) [reply polkadot/runtime/parachains/src/builder.rs]: fixed

**Comment by alindima** [comment]
> Not sure which benchmark you are refering to. If you re talking about the big inherent, it just doesn't make sense, that is for testing purposes, we don't need that weight.

A variation of the big inherent thingy. With realistic values, which exercises multiple candidates for a para (this is already allowed by the runtime). Otherwise, I'll add it in https://github.com/paritytech/polkadot-sdk/pull/3479 (but that PR is getting quite big and I could use the help :D )

File: polkadot/runtime/parachains/src/builder.rs (hunk 4 above)
Comment (alindima) [reply polkadot/runtime/parachains/src/builder.rs:686]: yeah, I was suggesting have this PR include these changes if useful.

Anyway, I don't think this'd be useful for benchmarking. And for tests, I don't think I can make a lot of use of the BenchBuilder (besides for a test or two, which is still useful).

I already added a lot of tests for paras_inherent (for candidate sanitization mostly) which exercise corner cases that are not easily expressable in such a high-level interface.

So it's fine to keep as is

**Review by alindima — APPROVED** [review]

**Review by eskimor — APPROVED** [review]
